### PR TITLE
Do not flow control headers

### LINKED ISSUE #4489: HTTP/2 headers should not be flow controlled (outbound)
We've discussed this in the past for a variety of reasons. Trailers definitely have to go through the flow controller since they can't be sent until after the last DATA frame. However, there isn't a good reason that I'm aware of that we shouldn't just send HEADERS immediately.

Adding HEADERS to outbound flow control will add work to the new `UniformByteDistributor` (#4464) since it has to keep track of zero-length frames separately ... and HEADERS are zero-length frames. 

As part of resolving this issue, we should refactor the `RST_STREAM` fix in #4502. We need a better solution for dealing with delayed `HEADERS` frames.

/cc @Scottmitch @louiscryan

### PR DESCRIPTION
**Pull request by nmittler**
Motivation:

Currently the encoder always sends headers to the flow controller. The fact that the headers are delayed can lead to issues such as RST_STREAM frames for streams unknown to the remote endpoint, as well as additional processing pressure on the remote flow controller.

Modifications:

Modified the encoder so that headers are only flow controlled if DATA has been sent on the stream.

Result:

Fixes #4489

### CODE REVIEW

**Comment by nmittler** [comment]
@Scottmitch @louiscryan PTAL

File: codec-http2/src/test/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoderTest.java [hunk 1]
```diff
@@ -617,22 +611,57 @@ public Void answer(InvocationOnMock invocationOnMock) throws Throwable {
...
+            public ChannelPromise addListener(
+                    GenericFutureListener<? extends Future<? super Void>> listener) {
+                throw new AssertionFailedError();
             }
-        }).when(future).addListener(any(ChannelFutureListener.class));
+
+            @Override
+            public ChannelPromise addListeners(
+                    GenericFutureListener<? extends Future<? super Void>>... listeners) {
+                throw new AssertionFailedError();
```
Comment (normanmaurer) [inline codec-http2/src/test/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoderTest.java]: `Assert.fail()` ?

File: codec-http2/src/test/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoderTest.java [hunk 2]
```diff
@@ -617,22 +611,57 @@ public Void answer(InvocationOnMock invocationOnMock) throws Throwable {
...
+
+            @Override
+            public boolean isVoid() {
+                return true;
+            }
+        };
+    }
+    private ChannelFuture newSucceededFuture() {
+        ChannelPromise future = newPromise();
+        future.setSuccess();
```
Comment (normanmaurer) [inline codec-http2/src/test/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoderTest.java]: `return future.setSuccess()`

File: codec-http2/src/test/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoderTest.java [hunk 3]
```diff
@@ -617,22 +611,57 @@ public Void answer(InvocationOnMock invocationOnMock) throws Throwable {
...
+    private Http2Stream reservePushStream(int pushStreamId, Http2Stream parent) throws Http2Exception {
+        return connection.local().reservePushStream(pushStreamId, parent);
+    }
+
+    private Http2Stream stream(int streamId) {
+        return connection.stream(streamId);
+    }
+
+    private void goAwayReceived(int lastStreamId) {
+        connection.goAwayReceived(lastStreamId, 0, Unpooled.EMPTY_BUFFER);
```
Comment (normanmaurer) [inline codec-http2/src/test/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoderTest.java]: `EMPTY_BUFFER` only ?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java [hunk 4]
```diff
@@ -355,7 +368,8 @@ public void write(ChannelHandlerContext ctx, int allowedBytes) {
 
         @Override
         public boolean merge(ChannelHandlerContext ctx, Http2RemoteFlowController.FlowControlled next) {
-            if (FlowControlledData.class != next.getClass()) {
+            if (FlowControlledData.class != next.getClass() ||
```
Comment (normanmaurer) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java]: btw can you tell me why we not just use "instanceof" ?

**Comment by normanmaurer** [comment]
@nmittler I know you not asked for my review but I could not resist :smile:

**Comment by nmittler** [comment]
As noted in #4489, closing this PR in favor of other modifications.

**Pull request closed without merging**